firewood/src/storage/linear/filebacked.rs

```diff
@@ -9,39 +9,61 @@
 // object. Instead, we probably should use an IO system that can perform multiple
 // read/write operations at once
 
-use super::{ReadLinearStore, WriteLinearStore};
-use std::fs::File;
+use std::fs::{File, OpenOptions};
 use std::io::{Error, Read, Seek};
 use std::os::unix::fs::FileExt;
 use std::path::PathBuf;
 use std::sync::Mutex;
 
 #[derive(Debug)]
-pub(super) struct FileBacked {
-    path: PathBuf,
+pub(crate) struct FileBacked {
     fd: Mutex<File>,
 }
 
-impl ReadLinearStore for FileBacked {
-    fn stream_from(&self, addr: u64) -> Result<Box<dyn Read>, Error> {
+impl FileBacked {
+    pub(super) fn new(path: PathBuf) -> Result<Self, Error> {
+        let fd = OpenOptions::new()
+            .read(true)
+            .write(true)
+            .create(true)
+            .truncate(false)
+            .open(path)?;
+
+        Ok(Self { fd: Mutex::new(fd) })
+    }
+
+    pub(super) fn stream_from(&self, addr: u64) -> Result<Box<dyn Read>, Error> {
         let mut fd = self.fd.lock().expect("p");
         fd.seek(std::io::SeekFrom::Start(addr))?;
         Ok(Box::new(fd.try_clone().expect("poisoned lock")))
     }
 
-    fn size(&self) -> Result<u64, Error> {
+    pub(super) fn size(&self) -> Result<u64, Error> {
         self.fd
             .lock()
             .expect("poisoned lock")
             .seek(std::io::SeekFrom::End(0))
     }
-}
 
-impl WriteLinearStore for FileBacked {
-    fn write(&mut self, offset: u64, object: &[u8]) -> Result<usize, Error> {
+    pub(super) fn write(&mut self, offset: u64, object: &[u8]) -> Result<usize, Error> {
         self.fd
             .lock()
             .expect("poisoned lock")
             .write_at(object, offset)
     }
 }
+
+#[cfg(test)]
+#[allow(clippy::unwrap_used)]
+pub(super) mod tests {
+    use super::*;
+    use tempdir::TempDir;
+
+    pub fn new_temp_filebacked(data: &[u8]) -> FileBacked {
+        let temp_dir = TempDir::new("test_historical_stream_from").unwrap();
+        let db_path = temp_dir.path().join("db");
+        let mut store = FileBacked::new(db_path).unwrap();
+        store.write(0, data).unwrap();
+        store
+    }
+}
```

Comment:
I've done this before, and it was definitely something I regretted. `tests` modules should never have anything other than private visibility. Having a function written in the parent module that's only callable from tests is perfectly fine. 
```suggestion
#[cfg(test)]
#[allow(clippy::unwrap_used)]
pub(super) fn new_temp_filebacked(data: &[u8]) -> FileBacked {
    let temp_dir = tempdir::TempDir::new("test_historical_stream_from").unwrap();
    let db_path = temp_dir.path().join("db");
    let mut store = FileBacked::new(db_path).unwrap();
    store.write(0, data).unwrap();
    store
}
```

Reply:
Alternatively, you could put `#[cfg(not(test))]` above `new` and make this function a static method.

Reply:
removed